File: backend/worker/handlers/watermark.py
from worker.handlers.base import BaseHandler


class WatermarkHandler(BaseHandler):
    async def execute(self, node_config, input_paths, output_path):
        video = input_paths["video"]
        image = input_paths["overlay"]
        position = node_config.get("position", "bottom_right")
        opacity = float(node_config.get("opacity", 0.8))
        scale = float(node_config.get("scale", 0.15))
        margin = int(node_config.get("margin", 10))

        # Scale watermark relative to video width
        scale_filter = f"[1:v]scale=iw*{scale}:-1,format=rgba,colorchannelmixer=aa={opacity}[wm]"

        # Position mapping
        positions = {
            "top_left": f"{margin}:{margin}",
            "top_right": f"W-w-{margin}:{margin}",
            "bottom_left": f"{margin}:H-h-{margin}",
            "bottom_right": f"W-w-{margin}:H-h-{margin}",
            "center": "(W-w)/2:(H-h)/2",
        }
        overlay_pos = positions.get(position, positions["bottom_right"])

        filter_complex = f"{scale_filter};[0:v][wm]overlay={overlay_pos}[v]"

        args = [
            "-i", video,
            "-i", image,
            "-filter_complex", filter_complex,
            "-map", "[v]", "-map", "0:a?",
            *self.build_video_encode_args("libx264", preset="fast", crf=23),
            "-c:a", "copy",
            output_path,
        ]
        await self.run_ffmpeg(args)

File: internal/worker/handlers/watermark.go
package handlers

import (
	"context"
	"errors"
	"strconv"

	vpffmpeg "github.com/Ctwqk/videoprocess/internal/worker/ffmpeg"
)

type WatermarkHandler struct {
	Runner vpffmpeg.Runner
}

func (h WatermarkHandler) NodeType() string {
	return "watermark"
}

func WatermarkArgs(videoPath string, overlayPath string, outputPath string, config map[string]any) []string {
	position := stringValue(config["position"], "bottom_right")
	opacity := floatValue(config["opacity"], 0.8)
	scale := floatValue(config["scale"], 0.15)
	margin := intValue(config["margin"], 10)

	overlayPos := map[string]string{
		"top_left":     strconv.Itoa(margin) + ":" + strconv.Itoa(margin),
		"top_right":    "W-w-" + strconv.Itoa(margin) + ":" + strconv.Itoa(margin),
		"bottom_left":  strconv.Itoa(margin) + ":H-h-" + strconv.Itoa(margin),
		"bottom_right": "W-w-" + strconv.Itoa(margin) + ":H-h-" + strconv.Itoa(margin),
		"center":       "(W-w)/2:(H-h)/2",
	}
	selected, ok := overlayPos[position]
	if !ok {
		selected = overlayPos["bottom_right"]
	}

	filterComplex := "[1:v]scale=iw*" + formatFloat(scale) + ":-1:flags=lanczos,format=rgba,colorchannelmixer=aa=" + formatFloat(opacity) + "[wm];[0:v][wm]overlay=" + selected + "[v]"
	args := []string{
		"-i", videoPath,
		"-i", overlayPath,
		"-filter_complex", filterComplex,
		"-map", "[v]",
		"-map", "0:a?",
	}
	args = append(args, intermediateVideoEncodeArgs("libx264")...)
	args = append(args, "-c:a", "copy", outputPath)
	return args
}

func (h WatermarkHandler) Args(videoPath, overlayPath, outputPath string, config map[string]any) []string {
	return WatermarkArgs(videoPath, overlayPath, outputPath, config)
}

func (h WatermarkHandler) Execute(ctx context.Context, inputPaths map[string]string, outputPath string, config map[string]any) (map[string]any, error) {
	videoPath := inputPaths["video"]
	if videoPath == "" {
		return nil, errors.New("missing input path on video port")
	}
	overlayPath := inputPaths["overlay"]
	if overlayPath == "" {
		return nil, errors.New("missing input path on overlay port")
	}
	if err := runFFmpeg(ctx, h.Runner, h.Args(videoPath, overlayPath, outputPath, config)); err != nil {
		return nil, err
	}
	return map[string]any{}, nil
}

func formatFloat(value float64) string {
	return strconv.FormatFloat(value, 'f', -1, 64)
}
